core-bundle/src/EventListener/DataContainer/DynamicPtableListener.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener\DataContainer;
+
+use Contao\Controller;
+use Contao\CoreBundle\DependencyInjection\Attribute\AsHook;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\Input;
+
+/**
+ * Sets the parent table for the current table, if enabled and not set.
+ *
+ * @internal
+ */
+#[AsHook('loadDataContainer', priority: 255)]
+class DynamicPtableListener
+{
+    public function __construct(private readonly ContaoFramework $framework)
+    {
+    }
+
+    public function __invoke(string $table): void
+    {
+        if (
+            !($GLOBALS['TL_DCA'][$table]['config']['dynamicPtable'] ?? null)
+            || !isset($GLOBALS['BE_MOD']) || isset($GLOBALS['TL_DCA'][$table]['config']['ptable'])
+        ) {
+            return;
+        }
+
+        if (!$do = $this->framework->getAdapter(Input::class)->get('do')) {
```

Comment:
We should use the request object here instead of the input class, don't you think?

Reply:
Unless we want to mark it as used or need another logic but that's just a guess

https://github.com/contao/contao/blob/7ee94937ec98315b068c1da0f995300cd3b86d08/core-bundle/contao/library/Contao/Input.php#L138

Reply:
Do we use the request anywhere else in our custom backend code? Pretty sure people also use `setGet` to fake option, dunno if our Input class would forward this to the request?
I basically kept the code exactly as it was except extracting it to a listener.

Reply:
> Do we use the request anywhere else in our custom backend code?

Yes, a lot. There is not a single instance of `Input` in the new code and I would be very much in favor of keeping it that way.

Reply:
> There is not a single instance of `Input` in the new code and I would be very much in favor of keeping it that way.

New code meaning anything in `src/`? Because we do still use `Contao\Input` there. e.g.: https://github.com/contao/contao/blob/fbef6d53f786a4f4bce06f207cb8dcc456bc6c00/core-bundle/src/EventListener/DataContainer/PreviewLinkListener.php#L72

Reply:
> There is not a single instance of `Input` in the new code

Well, sadly I was wrong about that 😢

Reply:
That should be adjusted though. Or is there any reason for that listener to use `Input`? /cc @ausi

Reply:
> Or is there any reason for that listener to use `Input`?

Only if you need “input encoding” or you need to respect changes made by `Input::setGet()`. Not sure if that is the case in the `PreviewLinkListener`? /cc @aschempp

Reply:
See #8680 for further discussion.